plugins/lookup/passwordstore.py

```diff
@@ -346,10 +345,8 @@ def check_pass(self):
                     if ':' in line:
                         name, value = line.split(':', 1)
                         self.passdict[name.strip()] = value.strip()
-            if (self.backend == 'gopass' or
-                    os.path.isfile(os.path.join(self.paramvals['directory'], self.passname + ".gpg"))
-                    or not self.is_real_pass()):
-                # When using real pass, only accept password as found if there is a .gpg file for it (might be a tree node otherwise)
+            if os.path.isfile(os.path.join(self.paramvals['directory'], self.passname + ".gpg")):
```

Comment:
This is why the unit test fails. You removed an important check to support pass wrapper scripts (that means, neither gopass nor pass). And that's exactly why we added a unit test there, so that such breakage would be detected.

Suggestion:
```suggestion
            if ((self.backend != 'gopass' and not self.is_real_pass()) or
                os.path.isfile(os.path.join(self.paramvals['directory'], self.passname + ".gpg"))):
```

Reply:
That definitely doesn't work:

```
❯ ansible-playbook playbook.yml 
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [localhost] *****************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************
ok: [localhost]

TASK [Print only first line] *****************************************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "Multi"
}

TASK [Print all lines] ***********************************************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "Multi\nline\nfile"
}

TASK [Print only first line (pass)] **********************************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "Multi"
}

TASK [Print all lines (pass)] ****************************************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n\n\n=\n \np\na\ns\ns\n:\n \nt\nh\ne\n \ns\nt\na\nn\nd\na\nr\nd\n \nu\nn\ni\nx\n \np\na\ns\ns\nw\no\nr\nd\n \nm\na\nn\na\ng\ne\nr\n \n=\n\n\n=\n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n=\n\n\n=\n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \nv\n1\n.\n7\n.\n3\n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n=\n\n\n=\n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n=\n\n\n=\n \n \n \n \n \n \n \n \n \n \n \n \n \nJ\na\ns\no\nn\n \nA\n.\n \nD\no\nn\ne\nn\nf\ne\nl\nd\n \n \n \n \n \n \n \n \n \n \n \n=\n\n\n=\n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \nJ\na\ns\no\nn\n@\nz\nx\n2\nc\n4\n.\nc\no\nm\n \n \n \n \n \n \n \n \n \n \n \n \n=\n\n\n=\n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n=\n\n\n=\n \n \n \n \n \n \nh\nt\nt\np\n:\n/\n/\nw\nw\nw\n.\np\na\ns\ns\nw\no\nr\nd\ns\nt\no\nr\ne\n.\no\nr\ng\n/\n \n \n \n \n \n \n \n=\n\n\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n=\n\n"
}

TASK [Print only first line (dir)] ***********************************************************************************************************
fatal: [localhost]: FAILED! => {"msg": "An unhandled exception occurred while running the lookup plugin 'community.general.passwordstore'. Error was a <class 'ansible.errors.AnsibleError'>, original message: passwordstore: passname dir not found and missing=error is set"}
...ignoring

TASK [Print all lines (dir)] *****************************************************************************************************************
fatal: [localhost]: FAILED! => {"msg": "An unhandled exception occurred while running the lookup plugin 'community.general.passwordstore'. Error was a <class 'ansible.errors.AnsibleError'>, original message: passwordstore: passname dir not found and missing=error is set"}
...ignoring

TASK [Print only first line (dir, pass)] *****************************************************************************************************
fatal: [localhost]: FAILED! => {"msg": "An unhandled exception occurred while running the lookup plugin 'community.general.passwordstore'. Error was a <class 'ansible.errors.AnsibleError'>, original message: passwordstore: passname dir not found and missing=error is set"}
...ignoring

TASK [Print all lines (dir, pass)] ***********************************************************************************************************
fatal: [localhost]: FAILED! => {"msg": "An unhandled exception occurred while running the lookup plugin 'community.general.passwordstore'. Error was a <class 'ansible.errors.AnsibleError'>, original message: passwordstore: passname dir not found and missing=error is set"}
...ignoring

PLAY RECAP ***********************************************************************************************************************************
localhost                  : ok=9    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=4   
```

I think I'll just refactor `is_real_pass` into something like `is_known_pass_implementation` which returns True for regular pass and gopass and False in other cases.

Reply:
Can you share the modified code?

Reply:
The logic is correct, the problem is a bug in "is_real_pass" (it's re-using self.passoutput). If you just change that to a local variable it all works ok.

Reply:
```diff
--- passwordstore.py~	2022-07-29 11:26:02.000000000 +0000
+++ passwordstore.py	2022-07-29 11:27:51.438206000 +0000
@@ -261,11 +261,11 @@
     def is_real_pass(self):
         if self.realpass is None:
             try:
-                self.passoutput = to_text(
+                passoutput = to_text(
                     check_output2([self.pass_cmd, "--version"], env=self.env),
                     errors='surrogate_or_strict'
                 )
-                self.realpass = 'pass: the standard unix password manager' in self.passoutput
+                self.realpass = 'pass: the standard unix password manager' in passoutput
             except (subprocess.CalledProcessError) as e:
                 raise AnsibleError(e)
```

Reply:
But the check should be the other way, so that users of pass won't have to pay for the extra invocation:

```suggestion
            if (os.path.isfile(os.path.join(self.paramvals['directory'], self.passname + ".gpg")) or
                (self.backend != 'gopass' and not self.is_real_pass())):
```

By the way, curly braces look like that `{ }` ;)

Reply:
It's literally just commit 7c81bce032e9281ff2a1f843e65f74047c68e633 with your patch applied: https://gist.github.com/TheLastProject/d4b866611cdf56c8f5ef474cf61d80d6

The behaviour makes no sense to me though, but it persists even after removing the `__pycache__` directory. I must be missing something, but I've spent the past 30 minutes applying and reverting the patch but it keeps causing this weird behaviour on:

```
- hosts: localhost
  tasks:
    - name: Print all lines (pass)
      ansible.builtin.debug:
        msg: "{{ lookup('community.general.passwordstore', 'dir/test returnall=true') }}"
      ignore_errors: True
```

Reply:
Oh, GitHub didn't update with your comments. I was responding to something old. Nvm my last comment.

Reply:
Removing the double negation makes it easier to read.

```suggestion
            if (os.path.isfile(os.path.join(self.paramvals['directory'], self.passname + ".gpg"))
                or not (self.backend == 'gopass' or self.is_real_pass())):
```

I ran your test playbook successfully. Can you try fetching/running my patched version of passwordstore.py, maybe there is a subtle difference:

```
curl https://gist.github.com/grembo/7c94b7da98d63b4b0ec6331ec9f50ca7/raw/2824eeb4cba10ce8d985a449ae0b03492a61061e/passwordstore.py
```